### DISCUSSION TITLE
[Design proposal] Create/edit datasets UI improvements

### DISCUSSION THREAD

**Discussion by ghost** [apache/superset#19476] [Ideas]
Hi all, 

In part three of the effort to improve the overall dataset experience ([overall flow proposal](https://github.com/apache/superset/discussions/18584) | [chart creation flow proposal](https://github.com/apache/superset/discussions/18803)), we're looking at UI improvements to the create/edit dataset experience. Please review key screens below for primary UI updates.

Note that these screens would be a full-screen experience accessible from the dataset CRUD view, and this UI would also be reused in a modal accessible from Explore as it is today. Also note these screens represent a physical dataset, and some UI tweaks will be made for virtual datasets.

1. **Create dataset** - For physical datasets, user can preview database table before selecting it to create a dataset. Virtual datasets are still created from SQL Lab as it is today.
<img width="1440" alt="create" src="https://user-images.githubusercontent.com/90715511/181155131-b1344318-a40a-41bc-a919-0e3e917162b9.png">

2. **Edit  / add columns** - User can select dataset column to open a sidebar and configure column settings. Note we are renaming "calculated column" to "virtual column" as that language resonated better with users in testing sessions.
<img width="1440" alt="edit-column" src="https://user-images.githubusercontent.com/90715511/181155336-a77373fb-b515-4dd9-a818-8704603080bc.png">

<img width="1440" alt="add-column" src="https://user-images.githubusercontent.com/90715511/181155292-c6ac4c1f-8ed7-491b-961a-43b33aa6cb6b.png">

3. **Edit / add metrics** - Similar to columns, user can select a metric to open a sidebar and configure metrics.
<img width="1440" alt="edit-metric" src="https://user-images.githubusercontent.com/90715511/181155505-1a3cedf4-8673-473c-9bb5-431c550ab835.png">

<img width="1440" alt="add-metric" src="https://user-images.githubusercontent.com/90715511/181155533-6722fab0-3d5f-480a-a164-008cffa1d7d8.png">

4. **Settings** - Dataset settings moved to a modal accessible from the dataset header.

![settings-basic](https://user-images.githubusercontent.com/90715511/181155610-a6d5e532-7537-4660-a3e3-5137b1d31a80.png)

![settings-source](https://user-images.githubusercontent.com/90715511/181155627-d353bad3-9ff6-4c9e-b943-98936581103d.png)

![settings-data](https://user-images.githubusercontent.com/90715511/181155637-9452e433-8920-4a42-abaf-8802c7bf6514.png)

![settings-metadata](https://user-images.githubusercontent.com/90715511/181155659-258531ff-3ec1-4aea-8e7f-1ff8f5033688.png)

![settings-query](https://user-images.githubusercontent.com/90715511/181155668-db69a206-590f-4595-9b9a-0a3b59b0c97d.png)

5. **Granular column syncing from source** - In a later phase, we plan to add functionality that will allow users to select which columns they would like to sync from their source to the dataset.
![sync-column](https://user-images.githubusercontent.com/90715511/181155895-db3950ba-e4b5-4ab7-90f6-3e2d37d3787e.png)

6. **Dataset usage** - In a later phase, we plan to add functionality to let users see where the dataset is being used.
![usage](https://user-images.githubusercontent.com/90715511/181155962-fc0bcda3-b333-4aab-a852-9fb314a85c2b.png)

Please let me know if you have any questions!

**Comment by michael-s-molina** [comment]
Hi @jess-dillard. Thank you for the work! I really like the improvements ✨ 

Some questions:

> 1. Create dataset - For physical datasets, user can preview database table before selecting it to create a dataset. Virtual datasets are still created from SQL Lab as it is today.

Do we really need the pagination for a preview or the first page would be sufficient?
If we do need pagination, are we going to show all pages? If not, how the user will know that there are only 3 pages because it's a preview?

> 2. Edit / add columns - User can select dataset column to open a sidebar and configure column settings. Note we are renaming "calculated column" to "virtual column" as that language resonated better with users in testing sessions.
> 3. Edit / add metrics - Similar to columns, user can select a metric to open a sidebar and configure metrics.

Love the sidebar to edit the properties. In regards to the actions icon, currently, Superset uses 3 vertical dots in most places. Are we changing them to horizontal dots? Personally, I prefer the vertical option because the horizontal one always reminds me of truncation.

Can we offer a search field for the columns and metrics tables?

**Reply by ghost** [reply]
Great questions! 

1. Maybe "preview" was the wrong word here - we do want users to be able to view _all_ database table columns here. TBD if this will be an engineering challenge - if it is, we can definitely consider limiting how many columns are viewable and adding some helper text to explain. 

2. I believe it's slightly more standard / visually appealing to use a horizontal ellipses icon in a table since the eye is tracking horizontally with the rows. We also use horizontal ellipses icons elsewhere in the product (e.g. in the chart / dashboard header to trigger the global actions menu), so this is consistent with that as well. @kasiazjc do you have opinions? We should decide now while we're building the table component :) 

3. When this was originally designed we weren't planning on adding a global search function to this table component, but that has changed since then. If we have global search ready to go in the component I have no issues using it here.

**Reply by michael-s-molina** [reply]
Thanks for the responses. Just to clarify a bit more:

> 1. Maybe "preview" was the wrong word here - we do want users to be able to view all database table columns here. TBD if this will be an engineering challenge - if it is, we can definitely consider limiting how many columns are viewable and adding some helper text to explain.

My question was more in the sense of table rows, not columns. Example: should we display only 20 rows as the sample and no pagination? All table rows using pagination? Only the first 3 pages of rows?

> 2. I believe it's slightly more standard / visually appealing to use a horizontal ellipses icon in a table since the eye is tracking horizontally with the rows. We also use horizontal ellipses icons elsewhere in the product (e.g. in the chart / dashboard header to trigger the global actions menu), so this is consistent with that as well. @kasiazjc do you have opinions? We should decide now while we're building the table component :)

We do have these icons mixed currently and both types open an options menu 😢

<img width="1781" alt="Screen Shot 2022-07-27 at 2 09 03 PM" src="https://user-images.githubusercontent.com/70410625/181308436-4cc9b1a7-5718-47e7-943e-aa4a06f16d72.png">

**Reply by ghost** [reply]
I think part of the confusion around the table is that it is a table of database table columns (each row is a column). I'm saying we would should all of the table columns, so there would be as much pagination as necessary to display all database table columns (as rows).

**Reply by michael-s-molina** [reply]
Hahahaha, I'm so sorry. I just noticed that! I think it's time for a coffee 😜 ☕

**Comment by motorqabd** [comment]
Where is the option to **rename a virtual dataset created from SQL Lab**? If one is available. 
If not, can raise an issue for the same.

Is there a 'rename dataset' option on clicking on the edit pencil under 'Actions' tab?
